linkerd/app/outbound/src/logical/router.rs

```diff
@@ -207,24 +207,22 @@ where
     M: Clone,
     K: Hash + Eq + Clone,
 {
-    fn update_backends(&mut self, target_addrs: &ahash::AHashSet<NameAddr>) -> bool {
+    fn update_backends(&mut self, addrs: &ahash::AHashSet<NameAddr>) -> bool {
         let removed = {
             let init = self.backends.len();
-            self.backends.retain(|addr, _| target_addrs.contains(addr));
+            self.backends.retain(|addr, _| addrs.contains(addr));
             init - self.backends.len()
         };
 
-        if target_addrs
-            .iter()
-            .all(|addr| self.backends.contains_key(addr))
-            // Don't return early if we still need to populate the default backend!
-            && !target_addrs.is_empty()
-        {
+        // We just removed all backends that aren't in the new addrs, so we
+        // we can skip further processing by comparing their lengths.
+        debug_assert!(addrs.len() >= self.backends.len());
+        if addrs.len() == self.backends.len() {
```

Comment:
this isn't quite right, i think; the tests i fixed in ccfcc364a43bca52dbed784b48ecf22542f36975 are failing again; i think because the initial route set needs to be generated and we'll indicate that it doesn't?

Reply:
oh nevermind, the test failure is in a stack test and it's only failing because the stack test doesn't make a profile that matches what we would generate from proto anymore